File: 1_javabase/src/main/java/算法java/排序_快速排序.java
package 算法java;

public class 排序_快速排序 {
    public static void main(String[] args) {

        int[] array = {-97,-2,-2,26,-365,88};
        int[] arr = fastSort(0,array,array.length-1);
        for (int  i = 0;i<arr.length;i++){
            System.out.println(arr[i]);
        }
    }

    public static int[] fastSort(int left,int[] array,int right){
        int l = left;
        int r = right;
        //先找到中轴
        int pivotValue = array[(left + right)/2];
        int temp = 0;
        //如果左边的index>右边的index的时候退出
        while (l < r){
            //从左边找到一个大于pivotValue的值
            while (array[l] <pivotValue){
                l++;
            }
            //从右边找到一个小于pivotValue的值
            while (array[r] >pivotValue){
                r--;
            }
            //优化如果l>=r表明左右分解完成
            if (l >= r ){
                break;
            }
            //开始交换
            temp = array[l];
            array[l] = array[r];
            array[r] = temp;
            //这两句代码的意思是 交换后的值 如果和中轴的值相同，那此次不用在比较了，肯定是要么比之前大，要么比之前小 直接比较下一次的值
            if (array[l] == pivotValue){
                r--;
            }
            if (array[r] == pivotValue){
                l++;
            }
        }
        if (l == r){
            l++;
            r--;
        }
        //左递归
        if (left < r){
            fastSort(left,array,r);
        }
        //右递归
        if (right > l){
            fastSort(l,array,right);
        }
        return array;

    }
}

File: 1_javabase/src/main/java/算法scala/排序_快速排序.scala
package 算法scala

import java.text.SimpleDateFormat
import util.control.Breaks._
import java.util
import java.util.Date
/*
数据结构和算法 -- 快速排序
  */
object QuickSort {
  def main(args: Array[String]): Unit = {

    val arr = Array(15,8,7,16,10)

//    val random = new util.Random()
//    val arr = new Array[Int](80000000)
//    for (i <- 0 until 80000000) { //循环的生成随机数
//      arr(i) = random.nextInt(8000000)
//    }
    println("排序前")
    val now: Date = new Date()
    val dateFormat: SimpleDateFormat =
      new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val date = dateFormat.format(now)
    println("排序前时间=" + date) //输出时间
    println(util.Arrays.toString(arr))

    quickSort(0, arr.length - 1, arr) // 调用快排
    println("排序后")

    val now2: Date = new Date()
    val date2 = dateFormat.format(now2)
    println("排序后时间=" + date2) //输出时间
    println(util.Arrays.toString(arr))
    //println(arr.mkString(" ")) // ( 0 )// -567 -9 0 23 78 70
  }

  /*
  1. left: 指定从数组的左边的下标 0
  2. right : 指定从数组的右边的下标 length -1
  3. arr : 进行排序的数组
   */
  def quickSort(left: Int, right: Int, arr: Array[Int]): Unit = {
    var l = left // 左下标
    var r = right // 右下标
    var pivot = arr((left + right) / 2) // 以中间的值为基准进行分割
    var temp = 0
    breakable {
      // while 语句的作用就是把比 pivot 小的数放到左边, 比pivot大的数放到右边
      while (l < r) {
        while (arr(l) < pivot) { //从左边找一个比 pivot 大的值对应下标
          l += 1
        }
        while (arr(r) > pivot) { //从右边找一个比 pivot 小的值对应下标
          r -= 1
        }
        if (l >= r) { // 说明本次交换结束,退出本次while
          break()
        }
        var temp = arr(l) //交换
        arr(l) = arr(r)
        arr(r) = temp
        //处理后,如果发现arr(l) == pivot 就r - =1 , 提高效率
        if (arr(l) == pivot) {
          r -= 1
        }
        //
        if ((arr(r)) == pivot) {
          l += 1
        }
      }
    }
    if (l == r) { // 提高效率
      l += 1
      r -= 1
    }

    if (left < r) { //向左递归排序
      quickSort(left, r, arr)
    }
    if (right > l) {
      //向右递归排序
      quickSort(l, right, arr)
    }
  }

}